fsm/fsm.h

```diff
@@ -11,27 +11,83 @@
 #ifndef FSM_H
 #define FSM_H
 
-#include "priority_queue.h"
-
 #include <inttypes.h>
+#include <stdbool.h>
+#include <stdlib.h>
+
+typedef struct fsm *fsm;
+
+typedef void (*state_function)(fsm);
+typedef void (*event_handler)(fsm, uint8_t event);
 
-typedef struct fsm fsm;
+typedef struct state {
+    event_handler handler;
+    state_function entry;
+    state_function exit;
+} fsm_state;
+
+/**
+ * @brief Initializes a FSM instance
+ * 
+ * @param state_count Number of states
+ * @param event_count Number of events
+ * @return fsm The initialized FSM handle
+ */
+fsm fsm_init(size_t state_count, size_t event_count);
 
-typedef uint16_t (*state_function)(fsm *);
+/**
+ * @brief Destroys the FSM
+ * 
+ * @param handle The FSM instance to deinit
+ */
+void fsm_deinit(fsm *handle);
 
-struct fsm {
-    uint16_t current_state;
-    uint16_t future_state;
-    uint16_t state_count;
+/**
+ * @brief Starts the FSM by running the entry function of the initial state
+ * 
+ * @param handle The FSM instance handle
+ */
+void fsm_start(fsm handle);
 
-    PQ_QueueTypeDef event_queue;
-    state_function **state_table;
-};
+/**
+ * @brief Adds a state to the internal state table
+ * 
+ * @param handle The FSM instance handle
+ * @param id ID of the state
+ * @param state FSM state definition
+ */
+void fsm_set_state(fsm handle, uint32_t id, fsm_state *state);
 
-void fsm_init(fsm *FSM, uint16_t state_count);
-void fsm_deinit(fsm *FSM);
-uint16_t fsm_get_state(fsm *FSM);
-void fsm_handle_event(fsm *FSM, uint16_t state);
-void fsm_run(fsm *FSM);
+/**
+ * @brief Executes a state transition
+ * @details Runs the exit state of the current state, and runs the entry state of the next state
+ * 
+ * @param handle The FSM instance handle
+ * @param state The next state
+ */
+void fsm_transition(fsm handle, uint32_t state);
+
+/**
+ * @brief Returns the current running state
+ * 
+ * @param handle The FSM instance handle
+ * @return uint32_t The current state
+ */
+uint32_t fsm_get_state(fsm handle);
+
+/**
+ * @brief Sets the internal register for a given event
+ * 
+ * @param handle The FSM instance handle
+ * @param event Event to set
+ */
+void fsm_catch_event(fsm handle, uint32_t event);
```

Comment:
What about renaming `fsm_catch_event` to `fsm_trigger_event` to better explicit its purpose?

Reply:
Will do